### PR TITLE
BTHome: New feature

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
Little weekend (or rather Thursday and Friday night) project of mine. This extends [Skald, who advertises to the world](https://doc.riot-os.org/group__ble__skald.html) with [BTHome](https://bthome.io/), "an energy efficient but flexible BLE format for devices to broadcast their sensor data and button presses". The main use case is to integrate BLE-powered devices into [Home Assistant](https://www.home-assistant.io/integrations/bthome).

Most of the spec is supported, except "Trigger based advertisements" (since Skald at the moment only advertises periodically). For encryption, I was able to confirm with the spec examples that at least from the generated output 
is the expected one, given the example input, but Home Assistant is for some reason not able to decrypt the message (even the one _in_ the examples)... So further investigation is needed. 

### Testing procedure
- [ ] TODO: Provide tests

So far only tested by flashing a node and seeing the device pop up in Home Assistant.

### Issues/PRs references
None

### CODE REVIEW

**Comment by miri64** [comment]
> Investigate why encryption does not work with HA

After talking to @BOZHENG001, I think the problem is that Skald is implementing an older BLE version than BTHome expects and thus fields (such as TXAdd which is used as part of the nonce for encryption) are not in the place Home Assistant expects them to be. Will try to confirm this somehow.

**Comment by miri64** [comment]
Just testing something @RIOT-OS/maintainers (will be removed after the test, so please ignore)

File: sys/net/ble/skald/skald.c [hunk 1]
```diff
@@ -122,14 +153,20 @@ void skald_adv_start(skald_ctx_t *ctx)
...
     ctx->cur_chan = 0;
     ctx->pkt.flags = (BLE_ADV_NONCON_IND | BLE_LL_FLAG_TXADD);
+#if IS_USED(MODULE_SKALD_UPDATE_PKT_CB)
+    event_queue_init(&ctx->queue);
+#endif
 
     /* start advertising */
     _sched_next(ctx);
+#if IS_USED(MODULE_SKALD_UPDATE_PKT_CB)
+    event_loop(&ctx->queue);
```
[the remark this answers is not in the document]
Comment (kfessel) [reply sys/net/ble/skald/skald.c:168]: instead of starting and running a queue exclusiv for skald a already exiting one might be used either given by parameter to `skald_adv_start` or `config`  or just depending on `event_theard`

[the remark this answers is not in the document]
Comment (mguetschow) [reply sys/net/ble/skald/skald.c:168]: > just depending on `event_thread`

That'd sound more sensible to me. Also, I think this dependence would then only be needed for `skald_bthome_saul` and not for `skald_update_pkt_cb`.

[the remark this answers is not in the document]
Comment (miri64) [reply sys/net/ble/skald/skald.c:168]: I think I just copied that behavior from the other skald “daemons” (that's what I see them as). I can have a look, if I find some time.

File: examples/skald_bthome/Makefile [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+USEMODULE += saul_default
+USEMODULE += skald_bthome_saul
+
+DEVELHELP ?= 1
+
+BTHOME_NAME = "RIOT"
+BTHOME_ADV_INTERVAL ?= 60000
+ENCRYPTION_KEY ?=
+
+ifneq ($(ENCRYPTION_KEY),)
```
Comment (crasbe) [inline examples/skald_bthome/Makefile]:
```suggestion
ifneq (,$(ENCRYPTION_KEY))
```

I think that's the more common syntax for comparisons with empty values.

File: examples/skald_bthome/main.c [hunk 3]
```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (C) 2024 Martine S. Lenders
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
```
Comment (crasbe) [inline examples/skald_bthome/main.c:4]:
```suggestion
/*
 * SPDX-FileCopyrightText: 2024 Martine S. Lenders
 * SPDX-License-Identifier: LGPL-2.1-only
 */
```

File: examples/skald_bthome/main.c [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+#include <errno.h>
+#include <stdio.h>
+
+#include "saul_reg.h"
+#include "ztimer.h"
+
+#include "net/skald/bthome.h"
+
+#ifndef CONFIG_BTHOME_SAUL_REG_DEVS
+#define CONFIG_BTHOME_SAUL_REG_DEVS     (16U)
```
Comment (crasbe) [inline examples/skald_bthome/main.c]:
```suggestion
#  define CONFIG_BTHOME_SAUL_REG_DEVS   (16U)
```

File: examples/skald_bthome/main.c [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+#endif
+
+#ifndef BTHOME_ADV_INTERVAL
+#define BTHOME_ADV_INTERVAL             (60000U)
```
Comment (crasbe) [inline examples/skald_bthome/main.c]:
```suggestion
#  define BTHOME_ADV_INTERVAL           (60000U)
```

File: examples/skald_bthome/main.c [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            offset = 'a' - 10;
+        }
+        else {
+            continue;
+        }
+        encryption_key[i / 2] |= (c - offset) << shift;
+        shift = (shift) ? 0 : 4;
+        key_len = (i / 2) + 1;
+        if (key_len > SKALD_BTHOME_KEY_LEN) {
+            printf("Key was too long: %u bytes\n", key_len);
```
Comment (crasbe) [inline examples/skald_bthome/main.c]: Perhaps you can print the maximum length here too.

File: examples/skald_bthome/main.c [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+            return -ENOBUFS;
+        }
+    }
+
+    if (key_len < SKALD_BTHOME_KEY_LEN) {
+            printf("Key was too short: %u bytes\n", key_len);
```
Comment (crasbe) [inline examples/skald_bthome/main.c]: Perhaps you can print the minimum length here too.

File: sys/include/net/skald.h [hunk 8]
```diff
@@ -46,10 +46,15 @@
 
 #include <stdint.h>
 
+#include "kernel_defines.h"
 #include "ztimer.h"
 #include "net/ble.h"
 #include "net/netdev/ble.h"
 
+#if IS_USED(MODULE_SKALD_UPDATE_PKT_CB)
+#include "event.h"
```
Comment (crasbe) [inline sys/include/net/skald.h]:
```suggestion
#  include "event.h"
```

File: sys/include/net/skald/bthome.h [hunk 9]
```diff
@@ -0,0 +1,490 @@
+/*
+ * Copyright (C) 2024 Martine S. Lenders
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
```
Comment (crasbe) [inline sys/include/net/skald/bthome.h:4]:
```suggestion
/*
 * SPDX-FileCopyrightText: 2024 Martine S. Lenders
 * SPDX-License-Identifier: LGPL-2.1-only
 */
```

File: sys/include/net/skald/bthome.h [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+
+#ifndef NET_SKALD_BTHOME_H
+#define NET_SKALD_BTHOME_H
```
Comment (crasbe) [inline sys/include/net/skald/bthome.h]:
```suggestion
#pragma once
```

File: sys/include/net/skald/bthome.h [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+ * @param[in] adv_itvl_ms   advertising interval in milliseconds
+ */
+void skald_bthome_advertise(skald_bthome_ctx_t *ctx, uint32_t adv_itvl_ms);
+
+#ifdef __cplusplus
+}
+#endif
+
+/** @} */
+#endif /* NET_SKALD_BTHOME_H */
```
Comment (crasbe) [inline sys/include/net/skald/bthome.h]:
```suggestion
```

File: sys/include/net/skald/bthome/defs.h [hunk 12]
```diff
@@ -0,0 +1,322 @@
+/*
+ * Copyright (C) 2024 Martine S. Lenders
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
```
Comment (crasbe) [inline sys/include/net/skald/bthome/defs.h:4]:
```suggestion
/*
 * SPDX-FileCopyrightText: 2024 Martine S. Lenders
 * SPDX-License-Identifier: LGPL-2.1-only
 */
```

File: sys/include/net/skald/bthome/defs.h [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+
+#ifndef NET_SKALD_BTHOME_DEFS_H
+#define NET_SKALD_BTHOME_DEFS_H
```
Comment (crasbe) [inline sys/include/net/skald/bthome/defs.h]:
```suggestion
#pragma once

```

File: sys/include/net/skald/bthome/defs.h [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+    BTHOME_BTN_LONG_TRIPLE_PRESS = 0x05,    /**< Long triple press */
+    BTHOME_BTN_HOLD_PRESS = 0x80,           /**< Hold press */
+} skald_bthome_btn_t;
+
+#ifdef __cplusplus
+}
+#endif
+
+/** @} */
+#endif /* NET_SKALD_BTHOME_DEFS_H */
```
Comment (crasbe) [inline sys/include/net/skald/bthome/defs.h]:
```suggestion
```

File: sys/net/ble/skald/skald.c [hunk 15]
```diff
@@ -46,6 +46,12 @@
 #define ADV_AA                  (0x8e89bed6)    /* access address */
 #define ADV_CRC                 (0x00555555)    /* CRC initializer */
 
+#if IS_USED(MODULE_SKALD_UPDATE_PKT_CB)
+#define _on_adv_evt_isr         _on_adv_evt_event
+#else
+#define _on_adv_evt_isr         _on_adv_evt
+#endif
```
Comment (crasbe) [inline sys/net/ble/skald/skald.c:53]:
```suggestion
#if IS_USED(MODULE_SKALD_UPDATE_PKT_CB)
#  define _on_adv_evt_isr       _on_adv_evt_event
#else
#  define _on_adv_evt_isr       _on_adv_evt
#endif
```

File: sys/net/ble/skald/skald_bthome.c [hunk 16]
```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright (C) 2024 Martine S. Lenders
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
```
Comment (crasbe) [inline sys/net/ble/skald/skald_bthome.c:4]:
```suggestion
/*
 * SPDX-FileCopyrightText: 2024 Martine S. Lenders
 * SPDX-License-Identifier: LGPL-2.1-only
 */
```

File: sys/net/ble/skald/skald_bthome.c [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+
+/**
+ * @{
+ *
+ * @file
+ * @author  Martine S. Lenders <mail@martine-lenders.eu>
+ */
```
Comment (crasbe) [inline sys/net/ble/skald/skald_bthome.c:11]: This is missing a `@brief` part.

File: sys/net/ble/skald/skald_bthome.c [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+
+#include "net/skald/bthome.h"
+
+#if IS_USED(MODULE_SKALD_BTHOME_ENCRYPT)
+#include "random.h"
+#include "crypto/modes/ccm.h"
+#endif
```
Comment (crasbe) [inline sys/net/ble/skald/skald_bthome.c]:
```suggestion
#if IS_USED(MODULE_SKALD_BTHOME_ENCRYPT)
#  include "random.h"
#  include "crypto/modes/ccm.h"
#endif
```

File: sys/net/ble/skald/skald_bthome_saul.c [hunk 19]
```diff
@@ -0,0 +1,678 @@
+/*
+ * Copyright (C) 2024 Martine S. Lenders
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
```
Comment (crasbe) [inline sys/net/ble/skald/skald_bthome_saul.c:4]:
```suggestion
/*
 * SPDX-FileCopyrightText: 2024 Martine S. Lenders
 * SPDX-License-Identifier: LGPL-2.1-only
 */
```

File: sys/net/ble/skald/skald_bthome_saul.c [hunk 20]
```diff
@@ -0,0 +1,678 @@
...
+
+/**
+ * @{
```
Comment (crasbe) [inline sys/net/ble/skald/skald_bthome_saul.c:7]: This header is missing a `@brief` command.

File: sys/include/net/skald/bthome.h [hunk 21]
```diff
@@ -0,0 +1,490 @@
...
+ *
+ * @author  Martine S. Lenders <mail@martine-lenders.eu>
+ */
+
+#include "byteorder.h"
+#include "kernel_defines.h"
+#include "net/skald.h"
+
+#if IS_USED(MODULE_SKALD_BTHOME_SAUL)
+#include "saul_reg.h"
```
Comment (crasbe) [inline sys/include/net/skald/bthome.h]:
```suggestion
#  include "saul_reg.h"
```

File: sys/include/net/skald/bthome.h [hunk 22]
```diff
@@ -0,0 +1,490 @@
...
+     * @brief   Copy of the SAUL registry entry.
+     *
+     * Must be initialized when calling @ref skald_bthome_saul_add()
+     * with saul_reg_t::next set to NULL.
+     */
+    saul_reg_t saul;
+    /**
+     * @brief   Object ID for the SAUL registry entry
+     *
+     * @see skald_bthome_id_t
```
Comment (crasbe) [inline sys/include/net/skald/bthome.h]: This creates a `See also` block, maybe you want to use `@ref`.

File: sys/include/net/skald/bthome.h [hunk 23]
```diff
@@ -0,0 +1,490 @@
...
+    uint8_t encrypt;
+#endif
+#if IS_USED(MODULE_SKALD_BTHOME_SAUL) || defined(DOXYGEN)
+    /**
+     * @brief   The index of the last device sent in skald_bthome_ctx_t::devs.
+     *
+     * Will be updated on each periodic advertisement to allow for fragmenting
+     * different measurement readings across multiple advertisements (in case all
+     * measurements from skald_bthome_ctx_t::devs are too large for one
+     * advertisement). Is initialized to 0.
```
Comment (crasbe) [inline sys/include/net/skald/bthome.h:148]: I don't know if it's stated somewhere else, but it would probably be a good idea to mention that fragmentation is not part of the official standard.

File: sys/include/net/skald/bthome.h [hunk 24]
```diff
@@ -0,0 +1,490 @@
...
+ *                              shortened name should be advertised. Must end with '\0' and may not
+ *                              make the length of the PDU exceed @ref NETDEV_BLE_PDU_MAXLEN
+ * @param[in] complete_name     The complete name for the BTHome advertisement. May be NULL, if no
+ *                              complete name should be advertised. Must end with '\0' and may not
+ *                              make the length of the PDU exceed @ref NETDEV_BLE_PDU_MAXLEN
+ * @param[in] dev_info          Flags for the device information. Currently, no flags are supported.
+ *
+ * @return The current size of the advertisement on success
+ * @return -EMSGSIZE, when @p shortened_name or @p complete_name cause the message to exceed
+ *         @ref NETDEV_BLE_PDU_MAXLEN. With NDEBUG=0 an assertion is raised in this case.
```
Comment (crasbe) [inline sys/include/net/skald/bthome.h]:
```suggestion
 * @return The current size of the advertisement on success or an error code on failure
 * @retval -EMSGSIZE when @p shortened_name or @p complete_name cause the message to exceed
 *         @ref NETDEV_BLE_PDU_MAXLEN. With NDEBUG=0 an assertion is raised in this case.
```

File: sys/include/net/skald/bthome.h [hunk 25]
```diff
@@ -0,0 +1,490 @@
...
+ *
+ * @param[in,out] ctx       The BTHome context. Must not be NULL and must be initialized with
+ *                          @ref skald_bthome_init().
+ * @param[in] obj_id        An object ID (see @ref skald_bthome_id_t).
+ * @param[in] data          The measurement data. Must not be NULL.
+ * @param[in] data_len      The length of @p data.
+ *
+ * @return The current size of the advertisement on success
+ * @return -EMSGSIZE, when @p data_len cause the message to exceed @ref NETDEV_BLE_PDU_MAXLEN.
+ *         With NDEBUG=0 an assertion is raised in this case.
```
Comment (crasbe) [inline sys/include/net/skald/bthome.h]:
```suggestion
 * @return The current size of the advertisement on success or an error code on failure
 * @retval -EMSGSIZE when @p data_len cause the message to exceed @ref NETDEV_BLE_PDU_MAXLEN.
 *         With NDEBUG=0 an assertion is raised in this case.
```

File: sys/include/net/skald/bthome.h [hunk 26]
```diff
@@ -0,0 +1,490 @@
...
+ * @pre @p data fits within the advertisement in @p ctx.
+ *
+ * @param[in,out] ctx       The BTHome context. Must not be NULL and must be initialized with
+ *                          @ref skald_bthome_init().
+ * @param[in] obj_id        An object ID (see @ref skald_bthome_id_t).
+ * @param[in] data          The measurement data.
+ *
+ * @return The current size of the advertisement on success
+ * @return -EMSGSIZE, when @p data cause the message to exceed @ref NETDEV_BLE_PDU_MAXLEN.
+ *         With NDEBUG=0 an assertion is raised in this case.
```
Comment (crasbe) [inline sys/include/net/skald/bthome.h]:
```suggestion
 * @return The current size of the advertisement on success or an error code on failure
 * @retval -EMSGSIZE when @p data cause the message to exceed @ref NETDEV_BLE_PDU_MAXLEN.
 *         With NDEBUG=0 an assertion is raised in this case.
```

File: sys/include/net/skald/bthome.h [hunk 27]
```diff
@@ -0,0 +1,490 @@
...
```
Comment (crasbe) [inline sys/include/net/skald/bthome.h]:
```suggestion
 * @return The current size of the advertisement on success or an error code on failure
 * @retval -EMSGSIZE when @p data cause the message to exceed @ref NETDEV_BLE_PDU_MAXLEN.
 *         With NDEBUG=0 an assertion is raised in this case.
```

File: sys/include/net/skald/bthome.h [hunk 28]
```diff
@@ -0,0 +1,490 @@
...
```
Comment (crasbe) [inline sys/include/net/skald/bthome.h]:
```suggestion
 * @return The current size of the advertisement on success or an error code on failure
 * @retval -EMSGSIZE when @p data cause the message to exceed @ref NETDEV_BLE_PDU_MAXLEN.
 *         With NDEBUG=0 an assertion is raised in this case.
```

File: sys/include/net/skald/bthome.h [hunk 29]
```diff
@@ -0,0 +1,490 @@
...
```
Comment (crasbe) [inline sys/include/net/skald/bthome.h]:
```suggestion
 * @return The current size of the advertisement on success or an error code on failure
 * @retval -EMSGSIZE when @p data cause the message to exceed @ref NETDEV_BLE_PDU_MAXLEN.
 *         With NDEBUG=0 an assertion is raised in this case.
```

File: sys/include/net/skald/bthome.h [hunk 30]
```diff
@@ -0,0 +1,490 @@
...
```
Comment (crasbe) [inline sys/include/net/skald/bthome.h]:
```suggestion
 * @return The current size of the advertisement on success or an error code on failure
 * @retval -EMSGSIZE when @p data cause the message to exceed @ref NETDEV_BLE_PDU_MAXLEN.
 *         With NDEBUG=0 an assertion is raised in this case.
```

File: sys/include/net/skald/bthome.h [hunk 31]
```diff
@@ -0,0 +1,490 @@
...
```
Comment (crasbe) [inline sys/include/net/skald/bthome.h]:
```suggestion
 * @return The current size of the advertisement on success or an error code on failure
 * @retval -EMSGSIZE when @p data cause the message to exceed @ref NETDEV_BLE_PDU_MAXLEN.
 *         With NDEBUG=0 an assertion is raised in this case.
```

File: sys/include/net/skald/bthome.h [hunk 32]
```diff
@@ -0,0 +1,490 @@
...
```
Comment (crasbe) [inline sys/include/net/skald/bthome.h]:
```suggestion
 * @return The current size of the advertisement on success or an error code on failure
 * @retval -EMSGSIZE when @p data cause the message to exceed @ref NETDEV_BLE_PDU_MAXLEN.
 *         With NDEBUG=0 an assertion is raised in this case.
```

File: sys/include/net/skald/bthome.h [hunk 33]
```diff
@@ -0,0 +1,490 @@
...
+    return skald_bthome_add_measurement(ctx, obj_id, &le_data, sizeof(data));
+}
+
+#if IS_USED(MODULE_SKALD_BTHOME_SAUL) || defined(DOXYGEN)
+/**
+ * @brief   Add SAUL registry entry to BTHome
+ *
+ * skald_bthome_saul_t::saul of @p saul must be a copy of what is found in @ref sys_saul_reg.
+ *
+ * @see @ref sys_saul_reg
```
Comment (crasbe) [inline sys/include/net/skald/bthome.h:419]: Since this adds a `See also` block, it would be nicer to have a little bit of text here too.

File: sys/include/net/skald/bthome.h [hunk 34]
```diff
@@ -0,0 +1,490 @@
...
+
+#if IS_USED(MODULE_SKALD_BTHOME_ENCRYPT) || defined(DOXYGEN)
+/**
+ * @brief   Encrypt the packet in the Skald context
+ *
+ * @param[in] ctx   A BTHome Skald context. Must not be NULL.
+ *
+ * @return  The total size of the encrypted packet on success.
+ * @return  -ENOBUFS, if the encrypted packet would not fit into a
+ *          BLE advertisement.
```
Comment (crasbe) [inline sys/include/net/skald/bthome.h]:
```suggestion
 * @return  The total size of the encrypted packet on success or an error code on failure
 * @retval  -ENOBUFS if the encrypted packet would not fit into a
 *          BLE advertisement.
```

File: sys/include/net/skald/bthome/defs.h [hunk 35]
```diff
@@ -0,0 +1,322 @@
...
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   BTHome object IDs
+ *
+ * _v_ is the value send, with the type in parenthesis.
```
Comment (crasbe) [inline sys/include/net/skald/bthome/defs.h]:
```suggestion
 * _v_ is the value sent, with the type in parenthesis.
```

File: sys/net/ble/skald/Makefile [hunk 36]
```diff
@@ -8,4 +8,16 @@ ifneq (,$(filter skald_eddystone,$(USEMODULE)))
...
+ifneq (,$(filter skald_bthome,$(USEMODULE)))
+  SRC += skald_bthome.c
+endif
+
+ifneq (,$(filter skald_bthome_saul,$(USEMODULE)))
+  SRC += skald_bthome_saul.c
+endif
+
+ifneq (,$(filter skald_bthome_encrypt,$(USEMODULE)))
+  $(shell $(COLOR_ECHO) "\n\n$(COLOR_RED)This currently does not work with Home Assistant!$(COLOR_RESET)\n\n" 1>&2)
```
Comment (crasbe) [inline sys/net/ble/skald/Makefile]: Should this just keep compiling or interrupt the compilation?

I got bonked from someone for using a red message for something that is not an error and does not terminate the compilation, everything that is just a warning should apparently be yellow 🤷

File: sys/net/ble/skald/skald_bthome.c [hunk 37]
```diff
@@ -0,0 +1,266 @@
...
+    uint8_t uuid[2];
+    uint8_t dev_info;
+} _bthome_data_hdr_t;
+
+/* flags: LE General Discoverable Mode | BR/EDR Not Supported */
+#define BTHOME_FLAGS { .hdr = { .length = 0x02, .ad_type = 0x01 }, 0x06 }
+
+static int _set_name(skald_ctx_t *ctx, const char *name, uint8_t offset, uint8_t type)
+{
+    if (name) {
```
Comment (crasbe) [inline sys/net/ble/skald/skald_bthome.c]: You could also check for `if (!name)` and return an error code, that would save indenting most of the code.

File: sys/net/ble/skald/skald_bthome_saul.c [hunk 38]
```diff
@@ -0,0 +1,678 @@
...
+                saul->obj_id = BTHOME_ID_TVOC;
+                saul->add_measurement = &_add_ppb_to_ugm3_uint16_measurement;
+                return 0;
+            default:
+                return -1;
+            }
+        default:
+            return -1;
+        }
+        return -1;
```
Comment (crasbe) [inline sys/net/ble/skald/skald_bthome_saul.c:296]:
```suggestion
```

Double return?

File: examples/skald_bthome/Makefile (hunk 2 above)
Comment (miri64) [reply examples/skald_bthome/Makefile]: grafik

**Comment by miri64** [comment]
Rebased to current master

File: sys/include/net/skald/bthome.h (hunk 23 above)
Comment (miri64) [reply sys/include/net/skald/bthome.h:148]: It is not, but the way it is implemented it is completely compliant with the "standard" (only thing that one needs to know is that a sensor value is then send not every _x_ intervals, but ever `#fragments` * _x_ intervals. But timing isn't part of the "standard" either.

File: sys/net/ble/skald/Makefile (hunk 36 above)
Comment (miri64) [reply sys/net/ble/skald/Makefile]: Then I make it yellow. It should continue compiling, since the encryption technically is correct, it just does not work with HA.

File: sys/net/ble/skald/skald_bthome.c (hunk 17 above)
Comment (miri64) [reply sys/net/ble/skald/skald_bthome.c:11]: For a c-file? That's news to me. It is not even really included into the doxygen.

Comment (crasbe) [reply sys/net/ble/skald/skald_bthome.c:11]: It's nice to have it complete 🤔

Comment (miri64) [reply sys/net/ble/skald/skald_bthome.c:11]: Isn't it already sufficiently complete :D?

Comment (miri64) [reply sys/net/ble/skald/skald_bthome.c:11]: I mean I would only write „Implementation of BTHome in Skald“ or something nonsensical like that. I don't see much value in that.

File: sys/net/ble/skald/skald_bthome_saul.c (hunk 20 above)
Comment (miri64) [reply sys/net/ble/skald/skald_bthome_saul.c:7]: See https://github.com/RIOT-OS/RIOT/pull/20996#discussion_r2324604674. This is not a header.

File: sys/include/net/skald/bthome.h (hunk 23 above)
Comment (miri64) [reply sys/include/net/skald/bthome.h:148]: (but that is basically what the text already states)

File: sys/include/net/skald/bthome.h (hunk 33 above)
Comment (miri64) [reply sys/include/net/skald/bthome.h:419]: Doxygen already replaces with with `SAUL Registry` which would be exactly the text I would add here ;-).

**Comment by miri64** [comment]
I think I addressed all your comments now.

What I will now do, to get this in a mergeable state:

- Split out encryption into another (draft) PR. Since we need BLE 5 support for skald for this, I don't expect that to be merged anytime soon, but this way we at least have a memorandum.
- Provide tests for BTHome and or BTHome+Skald.

**Comment by miri64** [comment]
Encryption part is now in https://github.com/RIOT-OS/RIOT/pull/21699

**Comment by miri64** [comment]
Added a test, it is still WIP though (did not check if the output is actually correct and I also want to increase the coverage).

File: tests/net/skald_bthome/main.c [hunk 39]
```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (C) 2025 TU Dresden
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
```
Comment (crasbe) [inline tests/net/skald_bthome/main.c:4]:
```suggestion
/*
 * SPDX-FileCopyrightText: 2024 TU Dresden
 * SPDX-License-Identifier: LGPL-2.1-only
 */
```

File: tests/net/skald_bthome/main.c [hunk 40]
```diff
@@ -0,0 +1,148 @@
...
+    )) < 0) {
+        printf("Unable to add humidity (factor 0.01, uint24) measurement: %s\n", strerror(res));
+        return 1;
+    }
+    // if ((res = skald_bthome_add_int8_measurement(
+    //     &_ctx, BTHOME_ID_TEMPERATURE_FACTOR_1, 0x12
+    // )) < 0) {
+    //     printf("Unable to add temperature factor 1 measurement: %s\n", strerror(res));
+    //     return 1;
+    // }
```
Comment (crasbe) [inline tests/net/skald_bthome/main.c:130]:
```suggestion
/*    if ((res = skald_bthome_add_int8_measurement(
 *        &_ctx, BTHOME_ID_TEMPERATURE_FACTOR_1, 0x12
 *    )) < 0) {
 *        printf("Unable to add temperature factor 1 measurement: %s\n", strerror(res));
 *        return 1;
 *    }
 */
```

File: tests/net/skald_bthome/tests/01-run.py [hunk 41]
```diff
@@ -0,0 +1,22 @@
+#! /usr/bin/env python3
+#
+# Copyright (C) 2025 TU Dresden
+#
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
```
Comment (crasbe) [inline tests/net/skald_bthome/tests/01-run.py]:
```suggestion
# SPDX-FileCopyrightText: 2025 TU Dresden
# SPDX-License-Identifier: LGPL-2.1-only
```

File: tests/net/skald_bthome/tests/01-run.py [hunk 42]
```diff
@@ -0,0 +1,22 @@
...
+
+import sys
+from testrunner import run
+
```
Comment (crasbe) [inline tests/net/skald_bthome/tests/01-run.py:8]:
```suggestion


```

To make the static test happy.

File: tests/net/skald_bthome/main.c [hunk 43]
```diff
@@ -0,0 +1,148 @@
...
+    (void)dev;
+    (void)max_len;
+
+    switch (opt) {
+        case NETOPT_DEVICE_TYPE:
+            assert(max_len >= sizeof(uint16_t));
+            *((uint16_t *)val) = NETDEV_TYPE_BLE;
+            return sizeof(uint16_t);
+        default:
+            return -ENOTSUP;
```
Comment (crasbe) [inline tests/net/skald_bthome/main.c]:
```suggestion
    case NETOPT_DEVICE_TYPE:
        assert(max_len >= sizeof(uint16_t));
        *((uint16_t *)val) = NETDEV_TYPE_BLE;
        return sizeof(uint16_t);
    default:
        return -ENOTSUP;
```

I think there is a coding convention to keep the case at the same indentation as the switch. This is not explicitly mentioned in the document, but can be seen here: https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md#indentation-of-preprocessor-directives

File: tests/net/skald_bthome/main.c [hunk 44]
```diff
@@ -0,0 +1,148 @@
...
+    (void)len;
+
+    switch (opt) {
+        case NETOPT_BLE_CTX:
+            return sizeof(netdev_ble_ctx_t);
+        case NETOPT_TX_POWER:
+            assert(len == sizeof(int16_t));
+            return sizeof(int16_t);
+        default:
+            return -ENOTSUP;
```
Comment (crasbe) [inline tests/net/skald_bthome/main.c]:
```suggestion
    case NETOPT_BLE_CTX:
        return sizeof(netdev_ble_ctx_t);
    case NETOPT_TX_POWER:
        assert(len == sizeof(int16_t));
        return sizeof(int16_t);
    default:
        return -ENOTSUP;
```

Same.

File: tests/net/skald_bthome/main.c (hunk 40 above)
Comment (miri64) [reply tests/net/skald_bthome/main.c:130]: I don't see any reason to keep this, actually (in case the test output is correct). However, it also serves as a test if the overflow protection works. So keeping as is ftm.
